rats-apps/src/python/rats/apps/_annotations.py

```diff
@@ -81,63 +97,103 @@ def get_namespaced_group(
-    return fn_annotation_decorator(ProviderNamespaces.GROUPS, group_id)
 
+class service(fn_annotation_decorator[T_ServiceType]):
+    def __init__(self, service_id: ServiceId[T_ServiceType]) -> None:
+        super().__init__(ProviderNamespaces.SERVICES, service_id)
 
-def config(
-    config_id: ConfigId[T_ConfigType],
-) -> Callable[[Callable[P, T_ConfigType]], Callable[P, T_ConfigType]]:
-    return fn_annotation_decorator(ProviderNamespaces.SERVICES, config_id)
 
+class group(fn_annotation_decorator[T_ServiceType]):
+    def __init__(self, group_id: ServiceId[T_ServiceType]) -> None:
+        super().__init__(ProviderNamespaces.GROUPS, group_id)
 
-def fallback_service(
-    service_id: ServiceId[T_ServiceType],
-) -> Callable[[Callable[P, T_ServiceType]], Callable[P, T_ServiceType]]:
-    return fn_annotation_decorator(ProviderNamespaces.FALLBACK_SERVICES, service_id)
 
+class config(fn_annotation_decorator[T_ConfigType]):
+    def __init__(self, config_id: ConfigId[T_ConfigType]) -> None:
+        super().__init__(ProviderNamespaces.SERVICES, config_id)
 
-def fallback_group(
-    group_id: ServiceId[T_ServiceType],
-) -> Callable[[Callable[P, T_ServiceType]], Callable[P, T_ServiceType]]:
-    return fn_annotation_decorator(ProviderNamespaces.FALLBACK_GROUPS, group_id)
 
+class fallback_service(fn_annotation_decorator[T_ServiceType]):
+    def __init__(self, service_id: ServiceId[T_ServiceType]) -> None:
+        super().__init__(ProviderNamespaces.FALLBACK_SERVICES, service_id)
 
-def fallback_config(
-    config_id: ConfigId[T_ConfigType],
-) -> Callable[[Callable[P, T_ConfigType]], Callable[P, T_ConfigType]]:
-    return fn_annotation_decorator(ProviderNamespaces.FALLBACK_SERVICES, config_id)
 
+class fallback_group(fn_annotation_decorator[T_ServiceType]):
+    def __init__(self, group_id: ServiceId[T_ServiceType]) -> None:
+        super().__init__(ProviderNamespaces.FALLBACK_GROUPS, group_id)
 
-def container(
-    group_id: ServiceId[T_ServiceType] = DEFAULT_CONTAINER_GROUP,
-) -> Callable[[Callable[P, T_ServiceType]], Callable[P, T_ServiceType]]:
-    return fn_annotation_decorator(ProviderNamespaces.CONTAINERS, group_id)
+
+class fallback_config(fn_annotation_decorator[T_ConfigType]):
+    def __init__(self, config_id: ConfigId[T_ConfigType]) -> None:
+        super().__init__(ProviderNamespaces.FALLBACK_SERVICES, config_id)
+
+
+class container(fn_annotation_decorator[T_ServiceType]):
+    def __init__(self, group_id: ServiceId[T_ServiceType] = DEFAULT_CONTAINER_GROUP) -> None:
+        super().__init__(ProviderNamespaces.CONTAINERS, group_id)
+
+
+@overload
+def autoid_service(
+    fn: Callable[[C], T_ServiceType],
+) -> Callable[[C, ServiceMethodArg], T_ServiceType]: ...
+
+
+@overload
+def autoid_service(
+    fn: Callable[[C, ServiceMethodArg], T_ServiceType],
+) -> Callable[[C, ServiceMethodArg], T_ServiceType]: ...
+
+
+def autoid_service(fn: Any) -> Any:
```

Comment:
is this functionally the same as a single function with a `Union` or the two types above, or is this falling back to allowing all types for `fn`? or is it something else?

Reply:
1. It's not falling back to allowing all types for `fn`.  Only calls to `autoid_service` that match one of the overloads will be allowed by the static analyser.
2. I did not try `Union` instead of overloading in this case.  But, before switching `service` to class, I tried the following, and it created problems with inferring `T_ServiceType`.  Switching to class + overloading solved them
```
def service(
    service_id: ServiceId[T_ServiceType],
) -> Callable[[Callable[[C], T_ServiceType] | Callable[[C, ServiceType], T_ServiceType]], Callable[P, T_ServiceType]]
```